### ISSUE TITLE
Element Plus prop values not typechecked

### ISSUE THREAD

**Issue by Ragura** [johnsoncodehk/volar#2176]
This may be a problem beyond the usage of Element Plus, but I will describe what happens for me:

When working inside a Vue template and using imported Element Plus components, there is no typechecking for values passed to a prop. It's strange, because when hovering over the component itself, it will correctly list the props and its accepted types.

For example: the `<el-image>` component has a prop called `fit` which can take the following values: `'fill' | 'contain' | 'cover' | 'none' | 'scale-down'`. The prop 'fit' correctly shows as available with intellisense (though I have to select it from the list, it won't autocomplete while typing inside the component tag). However, the prop itself is typed as `string` and has no mention of the correct list of strings as defined by the component.

<img width="892" alt="Screenshot 2022-12-07 at 14 35 02" src="https://user-images.githubusercontent.com/16363147/206193858-6927636f-4682-4b6a-8fd4-0fdcee016ca8.png">

<img width="892" alt="image" src="https://user-images.githubusercontent.com/16363147/206194048-a836bd8d-4ce9-4cd7-af67-ae123cc3597b.png">

Is there anything wrong with the type recognition, or did I set something up the wrong way?

**Comment by Sight-wcg** [comment]
I used the `Show Virtual Files` to look at the generated code, `SFCWithInstall`  causes type inference error.  

[element-plus: export type SFCWithInstall<T> = T & Plugin](https://github.com/element-plus/element-plus/blob/6aa69126b8fa491ad98006c60fa540b8843a7fd2/packages/utils/vue/typescript.ts#L3)
```js
// vue/runtime-core.d.ts
declare type Plugin_2 = (PluginInstallFunction & {
    install?: PluginInstallFunction;
}) | {
    install: PluginInstallFunction;
};
export { Plugin_2 as Plugin }

declare type PluginInstallFunction = (app: App, ...options: any[]) => any;

// VLS_type.ts
export type ComponentProps<T> =
	Record<string, unknown> &
	(
		T extends (...args: any) => any ? (T extends (...args: any) => { props: infer Props } ? Props : {})
		: T extends new (...args: any) => any ? (T extends new (...args: any) => { $props: infer Props } ? Props : {})
		: T // IntrinsicElement
	);
```
![image](https://user-images.githubusercontent.com/26325820/206411514-83ea93b0-68bb-4586-9fce-4069e8662950.png)

After removing `SFCWithInstall`, automatic completion can work normally.
![image](https://user-images.githubusercontent.com/26325820/206414536-a21a9fe2-1f61-4fc9-9404-cc8d3da9e9fb.png)

**Comment by johnsoncodehk** [comment]
@Sight-wcg Could you try does reverse `T extends new (...args: any) => any`, `T extends (...args: any) => any` fix the problem?
 
```ts
export type ComponentProps<T> =
	Record<string, unknown> &
	(
		T extends new (...args: any) => any ? (T extends new (...args: any) => { $props: infer Props } ? Props : {})
		: T extends (...args: any) => any ? (T extends (...args: any) => { props: infer Props } ? Props : {})
		: T // IntrinsicElement
	);
```

**Comment by johnsoncodehk** [comment]
FIxed by #2180

**Comment by Ragura** [comment]
Thank you for the quick resolution!
